{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "19ed4ddf-f8a8-4576-bde6-ac197d2d80bf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/miniconda3/envs/metric-full/lib/python3.11/site-packages/umap/__init__.py:9: ImportWarning: Tensorflow not installed; ParametricUMAP will be unavailable\n",
      "  warn(\n",
      "WARNING:root:In order to use the mouse gastrulation seqFISH datsets, please install squidpy (see https://github.com/scverse/squidpy).\n",
      "/root/miniconda3/envs/metric-full/lib/python3.11/site-packages/docrep/decorators.py:43: SyntaxWarning: 'param_categorical_covariate_keys' is not a valid key!\n",
      "  doc = func(self, args[0].__doc__, *args[1:], **kwargs)\n",
      "WARNING:root:In order to use sagenet models, please install pytorch geometric (see https://pytorch-geometric.readthedocs.io) and \n",
      " captum (see https://github.com/pytorch/captum).\n",
      "WARNING:root:mvTCR is not installed. To use mvTCR models, please install it first using \"pip install mvtcr\"\n",
      "WARNING:root:multigrate is not installed. To use multigrate models, please install it first using \"pip install multigrate\".\n",
      "/root/miniconda3/envs/metric-full/lib/python3.11/site-packages/scanpy/_settings.py:488: DeprecationWarning: `set_matplotlib_formats` is deprecated since IPython 7.23, directly use `matplotlib_inline.backend_inline.set_matplotlib_formats()`\n",
      "  IPython.display.set_matplotlib_formats(*ipython_format)\n"
     ]
    }
   ],
   "source": [
    "import scib\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from multiprocessing import Pool \n",
    "from scipy.io import mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "import muon\n",
    "import scarches as sca\n",
    "import scanpy as sc\n",
    "from scib_metrics.benchmark import Benchmarker\n",
    "import scib_metrics\n",
    "from typing import Optional\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad44d488-2fd9-4523-b592-a7ac1bc9d80c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dt_list = ['10_GSE201402_down']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6a122d42-a3f2-4165-86da-4c40f8599ec8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def read_RNA_ATAC(RNA_path,ATAC_path):\n",
    "    # gene expression\n",
    "    cell_names = pd.read_csv(RNA_path+'barcodes.tsv', sep = '\\t', header=None, index_col=None)\n",
    "    cell_names.columns =  ['cell_ids'] \n",
    "    cell_names['cell_ids'] = cell_names['cell_ids'].str.replace('.','-')\n",
    "    X = csr_matrix(mmread(RNA_path+'matrix.mtx').T)\n",
    "    gene_names = pd.read_csv(RNA_path+'features.tsv', sep = '\\t',  header=None, index_col=None) \n",
    "    gene_names.columns =  ['gene_ids'] \n",
    "    adata_RNA = ad.AnnData(X, obs=pd.DataFrame(index=cell_names.cell_ids), var=pd.DataFrame(index = gene_names.gene_ids))\n",
    "    adata_RNA.var_names_make_unique()\n",
    "    # peak information\n",
    "    cell_names = pd.read_csv(ATAC_path + 'barcodes.tsv', sep = '\\t', header=None, index_col=None)\n",
    "    cell_names.columns =  ['cell_ids'] \n",
    "    cell_names['cell_ids'] = cell_names['cell_ids'].str.replace('.','-')\n",
    "    X = csr_matrix(mmread(ATAC_path + 'matrix.mtx').T)\n",
    "    peak_name = pd.read_csv(ATAC_path + 'features.tsv', sep = '\\t',header=None,index_col=None)\n",
    "    peak_name.columns = ['peak_ids']\n",
    "    adata_ATAC  = ad.AnnData(X, obs=pd.DataFrame(index=cell_names.cell_ids), var=pd.DataFrame(index = peak_name.peak_ids))\n",
    "    return adata_RNA, adata_ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "34d6a584-bce1-41d8-925e-2ae15241fb77",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def count_metrics(dataset):\n",
    "    data_path = '/mnt/' + dataset\n",
    "    adata_RNA, adata_ATAC = read_RNA_ATAC(data_path+\"/RNA/\",data_path+\"/ATAC/\")\n",
    "\n",
    "    adata_RNA.layers[\"counts\"] = adata_RNA.X.copy()\n",
    "    sc.pp.normalize_total(adata_RNA)\n",
    "    sc.pp.log1p(adata_RNA)\n",
    "    sc.pp.highly_variable_genes(\n",
    "        adata_RNA,\n",
    "        flavor=\"seurat_v3\",\n",
    "        n_top_genes=4000,\n",
    "        subset=False\n",
    "    )\n",
    "    adata_RNA = adata_RNA[:, adata_RNA.var.highly_variable].copy()\n",
    "\n",
    "    adata_ATAC.layers['counts'] = adata_ATAC.X.copy()\n",
    "    sc.pp.normalize_total(adata_ATAC, target_sum=1e4)\n",
    "    sc.pp.log1p(adata_ATAC)\n",
    "    adata_ATAC.layers['log-norm'] = adata_ATAC.X.copy()\n",
    "    sc.pp.highly_variable_genes(adata_ATAC, n_top_genes=30000)\n",
    "    adata_ATAC = adata_ATAC[:, adata_ATAC.var.highly_variable].copy()\n",
    "    \n",
    "    # adata = sca.models.organize_multiome_anndatas(\n",
    "    #     adatas = [[adata_RNA], [adata_ATAC]],    # a list of anndata objects per modality, RNA-seq always goes first\n",
    "    #     layers = [['counts'], ['log-norm']], # if need to use data from .layers, if None use .X\n",
    "    # )\n",
    "    adata = sc.AnnData(obs=pd.DataFrame(index=adata_RNA.obs_names))\n",
    "    metadata = pd.read_csv(\"/mnt/\" + dataset + \"/metadata.csv\")\n",
    "    metadata['celltype'].index = adata.obs_names\n",
    "    adata.obs['cell_type'] = metadata['celltype'].astype('category')\n",
    "    if np.where(adata.obs[\"cell_type\"].isna())[0].shape[0]!=0:\n",
    "        adata.obs[\"cell_type\"] = adata.obs[\"cell_type\"].cat.add_categories(['NaN'])\n",
    "        adata.obs[\"cell_type\"][np.where(adata.obs[\"cell_type\"].isna())[0]] = 'NaN'\n",
    "    adata.obs['batch'] = ['batch1'] * adata.shape[0]\n",
    "    \n",
    "    result = pd.DataFrame()\n",
    "    metrics_list = []\n",
    "    method_list = ['scMVP','Schema']\n",
    "    for method in method_list:\n",
    "        if os.path.exists('/mnt/appealing/10_GSE201402_down/' + method + '.csv'):\n",
    "            latent = pd.read_csv('/mnt/appealing/10_GSE201402_down/'+ method + '.csv', header = None)\n",
    "            latent.index = adata.obs_names\n",
    "            adata.obsm[method] = latent\n",
    "            sc.pp.neighbors(adata, use_rep=method)\n",
    "            sc.tl.umap(adata)\n",
    "            sc.tl.leiden(adata, key_added=\"cluster\")\n",
    "            scib.metrics.cluster_optimal_resolution(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "            ari = scib.metrics.ari(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "            iso_asw = scib.metrics.isolated_labels_asw(adata, label_key=\"cell_type\", batch_key='batch', embed=method,  verbose = False)\n",
    "            nmi = scib.metrics.nmi(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "            # clisi = scib.metrics.clisi_graph(adata, label_key=\"cell_type\",use_rep=method, type_='embed')\n",
    "            sht = scib.metrics.silhouette(adata, label_key=\"cell_type\", embed=method, metric='euclidean', scale=True)\n",
    "            metrics_list.append([ari, iso_asw, nmi, sht, method])\n",
    "            \n",
    "    con = mmread('/mnt/appealing/10_GSE201402_down/' + 'Seurat' + '_connectivities.mtx')\n",
    "    dis = mmread('/mnt/appealing/10_GSE201402_down/' + 'Seurat' + '_distance.mtx')\n",
    "    adata.uns['neighbors'] = {'connectivities_key': 'connectivities', 'distances_key': 'distances', \n",
    "                              'params': {'n_neighbors': 20, 'method': 'umap', 'random_state': 0, \n",
    "                                         'metric': 'euclidean'}}\n",
    "    adata.uns['neighbors']['distance'] = csr_matrix(dis)\n",
    "    adata.uns['neighbors']['connectivities'] = csr_matrix(con)\n",
    "    adata.obsp['distance'] = csr_matrix(dis)\n",
    "    adata.obsp['connectivities'] = csr_matrix(con)\n",
    "    sc.tl.umap(adata, n_components=20)\n",
    "    scib.metrics.cluster_optimal_resolution(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "    ari = scib.metrics.ari(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "    iso_asw = scib.metrics.isolated_labels_asw(adata, label_key=\"cell_type\", batch_key='batch', embed=method,  verbose = False)\n",
    "    nmi = scib.metrics.nmi(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "    # clisi = scib.metrics.clisi_graph(adata, label_key=\"cell_type\",use_rep=method, type_='embed')\n",
    "    sht = scib.metrics.silhouette(adata, label_key=\"cell_type\", embed=method, metric='euclidean', scale=True)\n",
    "    metrics_list.append([ari, iso_asw, nmi, sht, 'Seurat'])\n",
    "\n",
    "    df = pd.DataFrame(metrics_list, columns=['ARI', 'ISO_ASW', 'NMI', 'Silhouette', 'method'])\n",
    "    result = df\n",
    "    result['Dataset'] = dataset\n",
    "    result.to_csv(data_path + \"/metrics_result1.csv\",index = False)\n",
    "    print(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d9813a1c-e4c4-40d3-8aa0-67050f72d0cd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: cluster key cluster already exists in adata.obs and will be overwritten because force=True \n",
      "resolution: 0.1, nmi: 0.8650851990842399\n",
      "resolution: 0.2, nmi: 0.886368427310364\n",
      "resolution: 0.3, nmi: 0.7616788267552257\n",
      "resolution: 0.4, nmi: 0.7433073581995274\n",
      "resolution: 0.5, nmi: 0.7401207335007871\n",
      "resolution: 0.6, nmi: 0.671553022305028\n",
      "resolution: 0.7, nmi: 0.6457632772713251\n",
      "resolution: 0.8, nmi: 0.6428685439184371\n",
      "resolution: 0.9, nmi: 0.6109411856141576\n",
      "resolution: 1.0, nmi: 0.5835742867058128\n",
      "resolution: 1.1, nmi: 0.56721866735733\n",
      "resolution: 1.2, nmi: 0.5579618390800895\n",
      "resolution: 1.3, nmi: 0.5472011904365633\n",
      "resolution: 1.4, nmi: 0.5597468616185625\n",
      "resolution: 1.5, nmi: 0.5403220387831961\n",
      "resolution: 1.6, nmi: 0.5344644824059293\n",
      "resolution: 1.7, nmi: 0.5244313255146976\n",
      "resolution: 1.8, nmi: 0.5263683387789307\n",
      "resolution: 1.9, nmi: 0.5179633982544485\n",
      "resolution: 2.0, nmi: 0.5229305114499204\n",
      "optimised clustering against cell_type\n",
      "optimal cluster resolution: 0.2\n",
      "optimal score: 0.886368427310364\n",
      "WARNING: cluster key cluster already exists in adata.obs and will be overwritten because force=True \n",
      "resolution: 0.1, nmi: 0.6983722434067615\n",
      "resolution: 0.2, nmi: 0.6868522948327669\n",
      "resolution: 0.3, nmi: 0.6608202791312737\n",
      "resolution: 0.4, nmi: 0.6095369009710211\n",
      "resolution: 0.5, nmi: 0.5895672635955802\n",
      "resolution: 0.6, nmi: 0.5559422712575044\n",
      "resolution: 0.7, nmi: 0.5739447943725112\n",
      "resolution: 0.8, nmi: 0.5648626502239564\n",
      "resolution: 0.9, nmi: 0.542764445195152\n",
      "resolution: 1.0, nmi: 0.531646985526425\n",
      "resolution: 1.1, nmi: 0.5316113504448827\n",
      "resolution: 1.2, nmi: 0.532575201413892\n",
      "resolution: 1.3, nmi: 0.514576527416326\n",
      "resolution: 1.4, nmi: 0.5078737668522552\n",
      "resolution: 1.5, nmi: 0.5160344470134706\n",
      "resolution: 1.6, nmi: 0.49549208009472134\n",
      "resolution: 1.7, nmi: 0.4954790250036508\n",
      "resolution: 1.8, nmi: 0.4929234945516519\n",
      "resolution: 1.9, nmi: 0.49071933310262106\n",
      "resolution: 2.0, nmi: 0.4847129822790692\n",
      "optimised clustering against cell_type\n",
      "optimal cluster resolution: 0.1\n",
      "optimal score: 0.6983722434067615\n",
      "WARNING: cluster key cluster already exists in adata.obs and will be overwritten because force=True \n",
      "resolution: 0.1, nmi: 0.7214015046997803\n",
      "resolution: 0.2, nmi: 0.6721367607655789\n",
      "resolution: 0.3, nmi: 0.6398970425061248\n",
      "resolution: 0.4, nmi: 0.6356832300412549\n",
      "resolution: 0.5, nmi: 0.614977712349117\n",
      "resolution: 0.6, nmi: 0.6178111816556907\n",
      "resolution: 0.7, nmi: 0.6161685780489682\n",
      "resolution: 0.8, nmi: 0.5822859577050908\n",
      "resolution: 0.9, nmi: 0.5815988961626214\n",
      "resolution: 1.0, nmi: 0.5762033125793318\n",
      "resolution: 1.1, nmi: 0.5807309659176974\n",
      "resolution: 1.2, nmi: 0.5805842470938566\n",
      "resolution: 1.3, nmi: 0.5555579742434649\n",
      "resolution: 1.4, nmi: 0.5639674540901689\n",
      "resolution: 1.5, nmi: 0.5498919681970152\n",
      "resolution: 1.6, nmi: 0.5520436130369303\n",
      "resolution: 1.7, nmi: 0.5450324092825244\n",
      "resolution: 1.8, nmi: 0.5401777990570513\n",
      "resolution: 1.9, nmi: 0.5370630278024701\n",
      "resolution: 2.0, nmi: 0.5372704316240463\n",
      "optimised clustering against cell_type\n",
      "optimal cluster resolution: 0.1\n",
      "optimal score: 0.7214015046997803\n",
      "10_GSE201402_down\n"
     ]
    }
   ],
   "source": [
    "for dataset in dt_list:\n",
    "    count_metrics(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80da8ba1-70fa-4f9c-8dc9-de12fa1c1dad",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "myconda",
   "language": "python",
   "name": "myconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
